content/into.md

```diff
@@ -0,0 +1,306 @@
+
+We can support implicit conversions to `Modifier`s simply by making `Modifier` an `into` trait:
+```scala
+into trait Modifier ...
+```
+This means implicit `Conversion` instances with `Modifier` results can be inserted without requiring a language import.
+
+Here is a simplified example:
+```scala
+trait Modifier
+given Conversion[Option[Node], Modifier] = ...
+given Conversion[Seq[Node], Modifier] = ...
+
+def f(x: Source, m: Modifier) = ...
+f(source, Some(node)) // inserts conversion
+```
+
+The `into`-as-a-modifier scheme is handy in codebases that have a small set of specific types that are intended as the targets of implicit conversions defined in the same codebase. Laminar's `Modifier` is a typical example. But the scheme can be easily abused by making the number of `into` types too large. One should restrict the number of `into`-declared types to the absolute minimum. In particular, never make a type `into` to just cater for the possibility that someone might want to later add an implicit conversion to it.
+
+
+## Details: Conversion target types
+
+To make the preceding descriptions more precise: An implicit conversion is permitted without an `implicitConversions` language import if the target type is a valid conversion target type. A valid conversion target type is one of the following:
+
+ - a type of the form `into[T]`,
+ - a reference `p.C` to a class or trait `C` that is declared with an `into` modifier,
+   which can also be followed by type arguments,
+ - a type alias of a valid conversion target type,
+ - a match type that reduces to a valid conversion target type,
+ - an annotated type `T @ann` where `T` is a valid conversion target type,
+ - a refined type `T {...}` where `T` is a valid conversion target type,
+ - a union `T | U` of two valid conversion target types `T` and `U`,
+ - an intersection `T & U` of two valid conversion target types `T` and `U`,
+ - an instance of a type parameter that is explicitly instantiated to a valid conversion target type.
+
+
+Type parameters that are not fully instantiated do not count as valid conversion target types. For instance, consider:
+
+```scala
+  trait Token
+  class Keyword(str: String)
+  given Conversion[String, Keyword] = KeyWord(_)
+
+  List[into[Keyword]]("if", "then", "else")
+```
+This type-checks since the target type of the list elements is the type parameter of the `List.apply` method which is explicitly instantiated to `into[Keyword]`. On the other hand, if we continue the example as follows we get an error:
+```scala
+  val ifKW: into[Keyword] = "if"
+  val ys: List[into[Keyword]] = List(ifKW, "then", "else")
+```
+Here, the type variable of `List.apply` is not explicitly instantiated
+when we check the `List(...)` arguments (it is just upper-bounded by the target type `into[Keyword]`). This is not enough to allow
+implicit conversions on the second and third arguments.
```

Comment:
Perhaps explain here in simpler terms *why* it is "not enough"? And perhaps give an example that show how to make it type check? Also you sometimes use "expected type" and sometimes "target type" - is there a reason for this? I think "expected type" is easier to understand without specific type inference knowledge, while "target typing" is a more advanced concept...

Reply:
I use "target" in "conversion target types" since its shorter than "expected types of conversions" and lends itself better to a formal definition.